cpp/include/cugraph/mtmg/detail/per_device_edgelist.hpp

```diff
@@ -0,0 +1,275 @@
+#include <raft/core/host_span.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+namespace cugraph {
+namespace mtmg {
+namespace detail {
+
+/**
+ * @brief An edgelist for each GPU
+ *
+ * Manages an edge list for edges associated with a particular GPU.  Multiple threads
+ * can call the append() method, possibly concurrently.  To avoid constantly copying
+ * when the buffers fill up, the class will create a device buffer containing a
+ * number of elements specified in the constructor.  When that device buffer is full
+ * we will create a new buffer.
+ *
+ * When we try and use the edgelist we will consolidate the buffers, since at that
+ * time we know the entire size required.
+ *
+ * Important note, the expectation is that this object will be used in two phases:
+ *  1) The append() method will be used to fill buffers with edges
+ *  2) The edges will be consumed to create a graph
+ *
+ * These two phases are expected to be disjoint.  The calling process is expected to
+ * manage some barrier so that all threads are guaranteed to be completed before changing
+ * phases.  If an append() call (part of the filling phase) overlaps with calls to
+ * finalize_buffer(), consolidate_and_shuffle(), get_src(), get_dst(), get_wgt(),
+ * get_edge_id() and get_edge_type() then the behavior is undefined (data might change
+ * in some non-deterministic way).
+ */
+template <typename vertex_t, typename weight_t, typename edge_t, typename edge_type_t>
+class per_device_edgelist_t {
+ public:
+  per_device_edgelist_t()                                        = delete;
+  per_device_edgelist_t(per_device_edgelist_t const&)            = delete;
+  per_device_edgelist_t& operator=(per_device_edgelist_t const&) = delete;
+  per_device_edgelist_t& operator=(per_device_edgelist_t&&)      = delete;
+
+  per_device_edgelist_t(cugraph::mtmg::handle_t const& handle,
+                        size_t device_buffer_size,
+                        bool use_weight,
+                        bool use_edge_id,
+                        bool use_edge_type)
+    : device_buffer_size_{device_buffer_size},
+      current_pos_{0},
+      src_{},
+      dst_{},
+      wgt_{std::nullopt},
+      edge_id_{std::nullopt},
+      edge_type_{std::nullopt}
+  {
+    if (use_weight) { wgt_ = std::make_optional(std::vector<rmm::device_uvector<weight_t>>()); }
+
+    if (use_edge_id) { edge_id_ = std::make_optional(std::vector<rmm::device_uvector<edge_t>>()); }
+
+    if (use_edge_type) {
+      edge_type_ = std::make_optional(std::vector<rmm::device_uvector<edge_type_t>>());
+    }
+
+    create_new_buffers(handle);
+  }
+
+  per_device_edgelist_t(per_device_edgelist_t&& other)
+    : device_buffer_size_{other.device_buffer_size_},
+      current_pos_{other.current_pos_},
+      src_{std::move(other.src_)},
+      dst_{std::move(other.dst_)},
+      wgt_{std::move(other.wgt_)},
+      edge_id_{std::move(other.edge_id_)},
+      edge_type_{std::move(other.edge_type_)}
```

Comment:
I guess `= default` will do the same.

Reply:
`std::mutex` (the lock) is not moveable.  This move constructor skips that and creates a new mutex.  That does affect thread safety (if you move the object while you're using it).  I'll add documentation about that.

Reply:
Gotcha, yeah, my bad, I thought we already have the move operator implemented with `= default` and thought why `= default` for the move operator but a custom code for the move constructor. Just find out that the move operator is actually deleted.